source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1999,59 +2000,139 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
+            std::copy(multibind_params.textures.samplers, multibind_params.textures.samplers + first_count.textures.count, smplrnames);
 
-            std::copy(multibind_params.textureImages.textures, multibind_params.textureImages.textures + first_count.textureImages.count, imgnames + first_count.textureImages.first);
-            std::copy(multibind_params.textureImages.formats, multibind_params.textureImages.formats + first_count.textureImages.count, imgformats + first_count.textureImages.first);
-        }
-        if (needRebind)
-        {
-            COpenGLExtensionHandler::extGlBindBuffersRange(GL_UNIFORM_BUFFER, 0u, MAX_UBO_BINDING_COUNT, ubonames, ubooffsets, ubosizes);
-            COpenGLExtensionHandler::extGlBindBuffersRange(GL_SHADER_STORAGE_BUFFER, 0u, MAX_SSBO_BINDING_COUNT, ssbonames, ssbooffsets, ssbosizes);
-            COpenGLExtensionHandler::extGlBindTextures(0u, MAX_TEXTURE_BINDING_COUNT, texnames, nullptr); //targets=nullptr: assuming ARB_multi_bind (or GL>4.4) is always available
-            COpenGLExtensionHandler::extGlBindSamplers(0u, MAX_TEXTURE_BINDING_COUNT, smplrnames);
-            COpenGLExtensionHandler::extGlBindImageTextures(0u, MAX_IMAGE_BINDING_COUNT, imgnames, imgformats);
+            assert(ARRAY_SIZE <= first_count.textureImages.count);
+            std::copy(multibind_params.textureImages.textures, multibind_params.textureImages.textures + first_count.textureImages.count, imgnames);
+
+
+            GLsizei maxIndex{};
+            GLsizei count{};
+            bool printWarnings =
+#ifdef _IRR_DEBUG
+                true;
+#else
+                false;
+#endif
+
+            maxIndex = first_count.ubos.first + first_count.ubos.count;
+#define CLAMP_COUNT(resname,limit,printstr) \
+            maxIndex = first_count.resname.first + first_count.resname.count;\
+            if (printWarnings)\
+                os::Printer::log(std::string("Warning: tried to bind unsupported amount of ") + #printstr + " descriptors!", ELL_WARNING);\
+            (first_count.resname.count - std::max(0, static_cast<int32_t>(maxIndex)-static_cast<int32_t>(limit)))
+
+            count = CLAMP_COUNT(ubos, COpenGLExtensionHandler::maxUBOBindings, UBO);
+            extGlBindBuffersRange(GL_UNIFORM_BUFFER, first_count.ubos.first, count, ubonames, ubooffsets, ubosizes);
+            count = CLAMP_COUNT(ssbos, COpenGLExtensionHandler::maxSSBOBindings, SSBO);
+            extGlBindBuffersRange(GL_SHADER_STORAGE_BUFFER, first_count.ssbos.first, count, ssbonames, ssbooffsets, ssbosizes);
+            count = CLAMP_COUNT(textures, COpenGLExtensionHandler::maxTextureBindings, texture); //TODO should use maxTextureBindingsCompute for compute
+            extGlBindTextures(first_count.textures.first, count, texnames, nullptr); //targets=nullptr: assuming ARB_multi_bind (or GL>4.4) is always available
```

Comment:
why would it be necessary to unbind completely before setting new samplers and textures

Reply:
@Crisspl request for comment.

Reply:
The use of `std::copy` and stack arrays is still pointless here.

Reply:
see https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r333204559

Reply:
resolved